idtools/idmap.go

```diff
@@ -0,0 +1,187 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+   Copyright The containerd Authors.
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+       http://www.apache.org/licenses/LICENSE-2.0
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+// Copied from https://github.com/containerd/containerd/blob/2ca3ff87255a4aa6b4244cb942033d45b6d44546/internal/userns/idmap.go
+
+/*
+   This file is copied and customized based on
+   https://github.com/moby/moby/blob/master/pkg/idtools/idtools.go
+*/
+
+package idtools
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+const invalidID = 1<<32 - 1
+
+var invalidUser = User{Uid: invalidID, Gid: invalidID}
+
+// User is a Uid and Gid pair of a user
+//
+//nolint:revive
+type User struct {
+	Uid uint32
+	Gid uint32
+}
+
+// IDMap contains the mappings of Uids and Gids.
+//
+//nolint:revive
+type IDMap struct {
+	UidMap []specs.LinuxIDMapping `json:"UidMap"`
+	GidMap []specs.LinuxIDMapping `json:"GidMap"`
+}
+
+func LoadIDMap(id string, labels map[string]string) (IDMap, error) {
+	var idmap IDMap
+	uidmapJSON, okUID := labels[snapshots.LabelSnapshotUIDMapping]
+	gidmapJSON, okGID := labels[snapshots.LabelSnapshotGIDMapping]
+	if okUID && okGID {
+		if err := json.Unmarshal([]byte(uidmapJSON), &idmap.UidMap); err != nil {
```

Comment:
the idmap are sent via marshalling using the idmap serializer, need to use the corresponding deserializer (imported from containerd pr)

Reference PR: https://github.com/containerd/containerd/pull/10722/files#diff-d4bc8cdbfad4af153a088d3594ef1c77ca112ef540c70a6089424bfbd359a98a

```
func (i *IDMap) Unmarshal(uidMap, gidMap string) error {
+       unmarshal := func(str string, fn func(m specs.LinuxIDMapping)) error {
+               if len(str) == 0 {
+                       return nil
+               }
+               for _, mapping := range strings.Split(str, ",") {
+                       m, err := deserializeLinuxIDMapping(mapping)
+                       if err != nil {
+                               return err
+                       }
+                       fn(m)
+               }
+               return nil
+       }
+       if err := unmarshal(uidMap, func(m specs.LinuxIDMapping) {
+               i.UidMap = append(i.UidMap, m)
+       }); err != nil {
+               return err
+       }
+
+       return unmarshal(gidMap, func(m specs.LinuxIDMapping) {
+               i.GidMap = append(i.GidMap, m)
+       })
+}
+
+func deserializeLinuxIDMapping(str string) (specs.LinuxIDMapping, error) {
+       var (
+               hostID, ctrID, length int64
+       )
+       _, err := fmt.Sscanf(str, "%d:%d:%d", &ctrID, &hostID, &length)
+       if err != nil {
+               return specs.LinuxIDMapping{}, fmt.Errorf("input value %s unparsable: %w", str, err)
+       }
+       if ctrID < 0 || ctrID >= invalidID || hostID < 0 || hostID >= invalidID || length < 0 || length >= invalidID {
+               return specs.LinuxIDMapping{}, fmt.Errorf("invalid mapping \"%s\"", str)
+       }
+       return specs.LinuxIDMapping{
+               ContainerID: uint32(ctrID),
+               HostID:      uint32(hostID),
+               Size:        uint32(length),
+       }, nil
+}
```